### PR TITLE
Added FilterPersister to customize the way filters are stored

### PR DESCRIPTION
**Pull request by yann-eugone**
I am targetting this branch, because this feature as a backward compatibility.

## Changelog

```markdown
### Added
- Added some `Sonata\AdminBundle\Filter\Persister\FilterPersisterInterface` to externalise filter persistence to a dedicated class
- Added some `Sonata\AdminBundle\Filter\Persister\SessionFilterPersister` to store filters to session (as done now)

### Changed
- Changed `Sonata\AdminBundle\Admin\AbstractAdmin::getFilterParameters` to use the configured filter persister component
```

## To do

- [x] Update the tests
- [x] Update the documentation
- [x] Add an upgrade note

## Subject

I had to persist filters accross sessions, so I decided to persist values to a database.
But as all the code remains in the AbstractAdmin, I had to redefine the whole method in my admin base class.

I did this PR so we can extend easily the way filters are persisted. We can also create more out-of-the-box filter persisters, such as : ORM, ODM, Redis...

### CODE REVIEW

File: Admin/AbstractAdmin.php [hunk 1]
```diff
@@ -190,11 +192,9 @@
     protected $label;
 
     /**
-     * Whether or not to persist the filters in the session.
-     *
-     * @var bool
+     * @var FilterPersisterInterface|bool
      */
-    protected $persistFilters = false;
+    protected $filterPersister;
```
Comment (OskarStark) [inline Admin/AbstractAdmin.php]: this is a BC break

**Comment by OskarStark** [comment]
If we can stay BC `3.x` is fine, otherwise we should deprecate the current logic and do this PR against master

File: Admin/AbstractAdmin.php [hunk 2]
```diff
@@ -1493,7 +1505,15 @@ public function getLabel()
...
     public function setPersistFilters($persist)
     {
-        $this->persistFilters = $persist;
+        $this->filterPersister = $persist;
+    }
+
+    /**
+     * @param FilterPersisterInterface $filterPersister
+     */
+    public function setFilterPersister($filterPersister)
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Please use type hinting here, and make it optional so that `null` can be passed.

File: Admin/AbstractAdmin.php [hunk 3]
```diff
@@ -190,11 +192,9 @@
     protected $label;
 
     /**
-     * Whether or not to persist the filters in the session.
-     *
-     * @var bool
+     * @var FilterPersisterInterface|bool
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: `FilterPersisterInterface|null`

File: Admin/AbstractAdmin.php [hunk 4]
```diff
@@ -766,16 +766,28 @@ public function getFilterParameters()
     {
         $parameters = array();
 
+        // if filter persister was configured with `true`, use the default persister (session)
+        if ($this->filterPersister === true) {
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: `null`

File: Admin/AbstractAdmin.php [hunk 5]
```diff
@@ -766,16 +766,28 @@ public function getFilterParameters()
...
+            $this->filterPersister = new SessionFilterPersister($this->request->getSession());
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Couldn't this be done with DI, BTW? If no filterPersister is set, inject this default service?

File: Admin/AbstractAdmin.php [hunk 6]
```diff
@@ -766,16 +766,28 @@ public function getFilterParameters()
...
         // build the values array
         if ($this->hasRequest()) {
             $filters = $this->request->query->get('filter', array());
 
-            // if persisting filters, save filters to session, or pull them out of session if no new filters set
-            if ($this->persistFilters) {
-                if ($filters == array() && $this->request->query->get('filters') != 'reset') {
-                    $filters = $this->request->getSession()->get($this->getCode().'.filter.parameters', array());
+            // if filter persistence is configured
+            if ($this->filterPersister instanceof FilterPersisterInterface) {
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: What else can it be?

File: DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 7]
```diff
@@ -310,7 +310,13 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
             $persistFilters = (bool) $container->getParameter('sonata.admin.configuration.filters.persist');
         }
 
-        $definition->addMethodCall('setPersistFilters', array($persistFilters));
+        // if persist filters is a boolean use session persister as default
+        // if persist filters is a string, use service reference instead
+        if (is_bool($persistFilters)) {
+            $definition->addMethodCall('setPersistFilters', array($persistFilters));
```
Comment (greg0ire) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: If `true`, inject default service,  if `false`, don't do anything.

File: Admin/AbstractAdmin.php (hunk 6 above)
Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: i'm not a big fan of `falsy` tests, i rather test type of my object

File: Admin/AbstractAdmin.php (hunk 5 above)
Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: my aim was to avoid deprecation, but yeah i can do it like it, but the boolean mode will have to be removed

File: Admin/AbstractAdmin.php (hunk 1 above)
Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: you're right... if we want to do it cleanly, we should remove this property

File: Admin/AbstractAdmin.php (hunk 5 above)
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Getting rid of the boolean mode is a good thing, b/c then it will be easier to reason on the type of `filterPersister`, and use native checks like type hinting.

File: Admin/AbstractAdmin.php (hunk 6 above)
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Not a big fan either, it was an honest question b/c I hadn't seen it could be boolean yet.

**Comment by yann-eugone** [comment]
Should I close this PR, and reopen onto the master branch, with service injection ?

**Comment by greg0ire** [comment]
> Should I close this PR, and reopen onto the master branch, with service injection ?

You mean it can't be BC?

File: Filter/Persister/SessionFilterPersister.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        $this->session->set($admin.'.filter.parameters', $filters);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function reset($admin)
+    {
+        $this->session->remove($admin.'.filter.parameters');
```
Comment (greg0ire) [inline Filter/Persister/SessionFilterPersister.php]: Please isolate the pattern building in a dedicated method.

File: Filter/Persister/FilterPersisterInterface.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * @return array The persisted filters
+     */
+    public function get($admin);
+
+    /**
+     * Set persisted filters for given admin.
+     *
+     * @param string $admin   The admin code
+     * @param array  $filters The filters to persist
```
Comment (greg0ire) [inline Filter/Persister/FilterPersisterInterface.php]: Please document what the array is supposed to look like (can it be multidimensional? What type are the keys, what are they?)

**Comment by yann-eugone** [comment]
@greg0ire I think so. We need to remove the `persistFilters` property (it may be confusing), and use service setter injection only.

this is a BC right ?

File: Filter/Persister/FilterPersisterInterface.php (hunk 9 above)
Comment (yann-eugone) [inline Filter/Persister/FilterPersisterInterface.php]: The structure is the same than the filter values. It depends on what filters you are using.

**Comment by greg0ire** [comment]
Removing a protected property is not BC. It's a BC-break and the property should be deprecated instead.

Can you try making the PR on 3.x? If we can't find a solution, then you'll have to rebase on master, but right now it's hard for me to imagine what would be a BC-break.

**Comment by yann-eugone** [comment]
Ok, I juste pushed a new version.

File: Admin/AbstractAdmin.php [hunk 10]
```diff
@@ -192,11 +193,22 @@
...
+     * @deprecated since 3.x, to be removed in 4.0.
      */
     protected $persistFilters = false;
 
     /**
+     * Component responsible for persisting filters.
+     *
+     * @var FilterPersisterInterface|null
+     */
+    protected $filterPersister;
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: How about making it private?

File: Admin/AbstractAdmin.php [hunk 11]
```diff
@@ -192,11 +193,22 @@
     /**
      * Whether or not to persist the filters in the session.
      *
+     * NEXT_MAJOR: remove this property
+     *
      * @var bool
+     *
+     * @deprecated since 3.x, to be removed in 4.0.
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: We could use magic to trigger deprecation notices here, but since it's the admin class, which is already big, and other properties could be deprecated, this could become a maintenance nightmare, so I'd say leave it like that.

File: Admin/AbstractAdmin.php (hunk 10 above)
Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: When a code for myself, everything is private, but here, other properties are protected, i just tried to code like the rest

Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Ok, good thinking, but please make it `private` anyway, we're trying to change things.

File: Filter/Persister/SessionFilterPersister.php [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Build the session key, under which the filter should be stored for given admin code.
+     *
+     * @param string $admin The admin code
+     *
+     * @return string The storage key
+     */
+    protected function buildStorageKey($admin)
```
Comment (greg0ire) [inline Filter/Persister/SessionFilterPersister.php]: `private` please

File: Filter/Persister/SessionFilterPersister.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+
+/**
+ * This filter persister is storing filters in session.
+ * This is the default behavior.
+ *
+ * @author Yann Eugoné <eugone.yann@gmail.com>
+ */
+class SessionFilterPersister implements FilterPersisterInterface
```
Comment (greg0ire) [inline Filter/Persister/SessionFilterPersister.php]: Make it `final` please

File: Filter/Persister/FilterPersisterInterface.php [hunk 14]
```diff
@@ -32,7 +32,16 @@ public function get($admin);
...
+     * @param array  $filters The filters to persist. Structure :
+     *                        {string filter field} => array(
+     *                          "type" => {int filter type},
+     *                          "value" => {mixed filter value},
+     *                        ),
+     *                        ...,
+     *                        "_page" => {int page num},
+     *                        "_sort_by" => {string sort property},
+     *                        "_sort_order" => {string sort order (ASC|DESC)},
+     *                        "_per_page" => {int count rows per page}
```
Comment (greg0ire) [inline Filter/Persister/FilterPersisterInterface.php]: Awesome!

**Comment by greg0ire** [comment]
Looks BC to me.

File: DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 15]
```diff
@@ -304,13 +304,22 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
...
         if (isset($attributes['persist_filters'])) {
-            $persistFilters = (bool) $attributes['persist_filters'];
-        } else {
-            $persistFilters = (bool) $container->getParameter('sonata.admin.configuration.filters.persist');
+            $persistFilters = $attributes['persist_filters'];
         }
 
-        $definition->addMethodCall('setPersistFilters', array($persistFilters));
+        // if boolean config value provided, use the default persister : session
+        if (is_bool($persistFilters) && $persistFilters) {
```
Comment (jordisala1991) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Do we have to check if `is_bool($persistFilters)`. Can't we instead just check normally? This property is set from container parameter or from the attributes, and we can cast it previously on the other case.

Maybe we can change from scalarNode to booleanNode on the Bundle Configuration?

Just to have properties that we know they are a concrete type and can't be other things.

**Comment by jordisala1991** [comment]
This is super nice!

We should remember to update other bundles mocking some classes that extends AbstractAdmin, because those test will broke.

File: DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php (hunk 15 above)
Comment (yann-eugone) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: this code may also receive value from the tag of an admin service. I think that allowing `true` to be provided, may be a short but efficient way to configure the persistence.

**Comment by greg0ire** [comment]
> We should remember to update other bundles mocking some classes that extends AbstractAdmin, because those test will broke.

@OskarStark see? Don't mock what you don't own ;)

**Comment by jordisala1991** [comment]
Agree with that, but when you are extending a class that you dont own. Either you mock it or you dont test your logic. Ideally extend classes is not a good practice in TDD but in some cases like sonataMedia we are not TDDing, but just adding test for the current code. If I have to choose between fragile test or fragile code IMO prefer to fix test all day long.

**Comment by OskarStark** [comment]
> @OskarStark see? Don't mock what you don't own ;)

👍

**Comment by OskarStark** [comment]
@yann-eugone can you please fix StyleCi?

File: DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php (hunk 15 above)
Comment (jordisala1991) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: got it. So if it is a boolean and true, we just change its value to the default persister, otherwise we chech if the value is there set by config (other provider or default) and then we call to setFilterPersister. It is something like that? was confused at first

Comment (yann-eugone) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: you got it. configuring `persist_filters` to `true` means "i don't care how you persist it, but persist it" :)

**Comment by yann-eugone** [comment]
@OskarStark  just pushed the fix, but not sure to understand the problem with `AbstractAdmin::$filter` property

**Comment by jordisala1991** [comment]
StyleCI complains about the other of the properties. `$filterPersister` should be below of `$persistFilters`

**Comment by yann-eugone** [comment]
@OskarStark just pushed the fix for StyleCI

**Comment by yann-eugone** [comment]
@OskarStark @greg0ire ping. what's missing for this to be merged ?

File: Admin/AbstractAdmin.php [hunk 16]
```diff
@@ -1490,13 +1509,28 @@ public function getLabel()
...
      */
     public function setPersistFilters($persist)
     {
+        @trigger_error(
+            'The '.__METHOD__.' method is deprecated since version 3.x and will be removed in 4.0.',
+            E_USER_DEPRECATED
+        );
+
         $this->persistFilters = $persist;
```
Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: In my first try, this method has to work.
Now this method will just do nothing.

Comment (core23) [inline Admin/AbstractAdmin.php]: Please add a BC layer for this method

Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: What do you mean ?

Comment (core23) [inline Admin/AbstractAdmin.php]: If someone is still using this method to set (dis)allow persisting the filters, will this still work with your new solution?

Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: This method was kept to trigger deprecation, but no matter if you call it or not, the property behind it is not used anymore in this version.

Comment (core23) [inline Admin/AbstractAdmin.php]: So a feature won't work after this commit, if you're still using the old method?

Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: Yes, calling this method directly will do nothing.

**Comment by yann-eugone** [comment]
Hello team,

Any update on this ? Not sure what you are expecting of me with this. Could you please provide details on that. I'll be glad to push more.

**Comment by greg0ire** [comment]
@OskarStark why did you require an upgrade note? @yann-eugone docs are missing aren't they?

**Comment by yann-eugone** [comment]
@greg0ire you're right, I did not write any doc about this. I'll wait the code approval to do so.

**Comment by greg0ire** [comment]
I approved the code. Also, you should consider doing things in the other way around, because it is less painful: create the documentation first, because it is easier to write, and easier for use to understand and approve or reject. If we reject it, no time will have been spent writing code.

**Comment by yann-eugone** [comment]
I will not start any debate on how things have to be done... Sorry.

So the only thing left is : writing some docs about it. I believe updating configuration reference + an entry in the cookbook is enough ?

**Comment by greg0ire** [comment]
> I will not enter any debate on how things have to be done... Sorry.

It was just a suggestion to avoid pain, you do of course as you feel is the best.

> So the only thing left is : writing some docs about it. I believe updating configuration reference + an entry in the cookbook is enough ?

It sure would. A cookbook entry describing how to implement and use your own `FilterPersister` would be great, along with something showing how to tweak configuration to use the one you provided.

**Comment by yann-eugone** [comment]
First draft of cookbook entry pushed : https://github.com/sonata-project/SonataAdminBundle/pull/4331/commits/b759608dad60c32c866e9cc7535e0e812473bf5e

File: Resources/doc/cookbook/recipe_persisting_filters.rst [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+            persist_filters: true
+
+
+Choose the persistence strategy
+-------------------------------
+
+When you enable the filters persistence by setting ``persist_filters`` to ``true``.
+SonataAdmin will use the default filter persister : ``Sonata\AdminBundle\Filter\Persister\SessionFilterPersister`` (which is, by now, the only one provided).
+
+You can implement your own filter persister by creating a new class that implement the ``Sonata\AdminBundle\Filter\Persister\FilterPersisterInterface`` interface and registering it as a service.
```
Comment (greg0ire) [inline Resources/doc/cookbook/recipe_persisting_filters.rst]: - "implements" with an "s" please
- Also, please comply with the sf docs standards as much as possible:

> [Each line should break approximately after the first word that crosses the 72nd character (so most lines end up being 72-78 characters);](http://symfony.com/doc/current/contributing/documentation/standards.html)

File: Resources/doc/cookbook/recipe_persisting_filters.rst [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+    .. code-block:: yaml
+
+        # src/AppBundle/Resources/config/admin.yml
+
+        services:
+            app.admin.car:
+                class: AppBundle\Admin\CarAdmin
+                tags:
+                    - { name: sonata.admin, manager_type: orm, group: Demo, label: Car, persist_filters: filter_persister_service_id }
```
Comment (greg0ire) [inline Resources/doc/cookbook/recipe_persisting_filters.rst]: Please use the multiline syntax:

```
-
    name: sonata.admin
    manager_type: orm
    group: Demo
    label: Car
    persist_filters: filter_persister_service_id
```

**Comment by greg0ire** [comment]
@OskarStark I removed the requirement for an upgrade note. @sonata-project/contributors please review again

**Comment by jlamur** [comment]
What do you think about using another parameter to define the persist filter service id?

If I understand well, currently, the default persister is dynamically set when the `sonata_admin.persist_filters` parameter to `true`.

Introducing a new parameter would let, for example, the default service id be dumped by `config:dump-reference`.

**Comment by yann-eugone** [comment]
I agree with this, I hate parameters that could be eitheir boolean, string, etc... it is not very easy to understand.
But the initial tought was to avoid breaking things, so I did the less changes I could to get the things working.
I could reconsider this approach, but the feature may break more things.

**Comment by greg0ire** [comment]
I don't think you could break other things by adding new configuration parameters, could you?

**Comment by yann-eugone** [comment]
I may misunderstood what @jlamur asked.
I could add a new parameter in the configuration that will contains the default service id used for persisting filters.
This parameter should allow to be overrided per-admin so we will have two parameter per admin too.

Is it what you was thinking about ?

**Comment by yann-eugone** [comment]
What about https://github.com/sonata-project/SonataAdminBundle/pull/4331/commits/6adc6196c7c3f0d61cea14811a2c1cfe8b8055b8 ?

File: Admin/AbstractAdmin.php (hunk 16 above)
Comment (core23) [inline Admin/AbstractAdmin.php]: Can you find a BC way for this? Maybe set the `filterPersister` to null or to a `FilterPersisterInterface` instance

Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: I'm not sure to understand what you want for this method.

Are asking me to 

- set `$this->filterPersister` to `null` if `$persist` = `false`
- set `$this->filterPersister` to `new SessionFilterPersister()` if `$persist` = `true`

?

Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: see f10b82d155a5de73ff4b79613b23f66b81a8a92a

File: Admin/AbstractAdmin.php [hunk 19]
```diff
@@ -1531,6 +1532,23 @@ public function setPersistFilters($persist)
...
+                $request = $this->getRequest();
+                $session = $request->getSession();
+
+                if ($session === null) {
+                    throw new \RuntimeException('The Request object has not session available');
+                }
+
+                $this->filterPersister = new SessionFilterPersister($session);
+            } catch (\RuntimeException $exception) {
+                // no request nor session
```
Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Throwing then catching and not doing anything in the same method is a code smell, please don't do it.

Comment (yann-eugone) [inline Admin/AbstractAdmin.php]: So you accept that an exception will be thrown if this method is called when there is no request nor session ? (ie dependency injection for instance)

Comment (greg0ire) [inline Admin/AbstractAdmin.php]: Since the container is dumped after this method is called, that would create something useless so yeah, I think an exception should be thrown when this method is misused. Or at least the catch should contain a deprecation notice saying an exception will be thrown in the future, if you think this could break existing setups.

File: Admin/AbstractAdmin.php (hunk 16 above)
Comment (jlamur) [inline Admin/AbstractAdmin.php]: Still not BC because it worked before when the session wasn't set on the admin.

Anyway, I think this should work with the old content...
Just add a check for `$this->persistFilters !== false` [here](https://github.com/yann-eugone/SonataAdminBundle/blob/c7d317e9f31180ce523409d8692937d3512197f0/Admin/AbstractAdmin.php#L788) and don't forget to add a `NEXT_MAJOR` comment too.

File: Filter/Persister/SessionFilterPersister.php [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+
+/**
+ * This filter persister is storing filters in session.
+ * This is the default behavior.
+ *
+ * @author Yann Eugoné <eugone.yann@gmail.com>
+ */
+final class SessionFilterPersister implements FilterPersisterInterface
```
Comment (greg0ire) [inline Filter/Persister/SessionFilterPersister.php]: @jlamur probably because https://ocramius.github.io/blog/when-to-declare-classes-final/
We are currently trying to close the API to avoid having to worry about extension BC-breaks, but also for reasons mentioned in this article: http://fabien.potencier.org/pragmatism-over-theory-protected-vs-private.html

Comment (jlamur) [inline Filter/Persister/SessionFilterPersister.php]: Alright! It would be good to mention this in the `CONTRIBUTING.md` or somewhere else ;)

Comment (greg0ire) [inline Filter/Persister/SessionFilterPersister.php]: Indeed!

File: DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 21]
```diff
@@ -317,13 +317,20 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
...
-            $persistFilters = (bool) $attributes['persist_filters'];
-        } else {
-            $persistFilters = (bool) $container->getParameter('sonata.admin.configuration.filters.persist');
+            $persistFilters = $attributes['persist_filters'];
+        }
+        $filtersPersister = $container->getParameter('sonata.admin.configuration.filters.persister');
+        // override default configuration with admin config if set
+        if (isset($attributes['filters_persister'])) {
+            $filtersPersister = $attributes['filters_persister'];
+        }
```
Comment (yann-eugone) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Don't you want to do this in a dedicated pull request, for ALL the settings that share that behavior ?

**Comment by yann-eugone** [comment]
@greg0ire @jlamur @OskarStark @jordisala1991 @core23 
do you still want this feature or not ?
what am I supposed to do to get this merged if so ?

**Comment by greg0ire** [comment]
I think it's a great feature. To get it merged, you need to resolve conflicts.

**Comment by greg0ire** [comment]
Please write a changelog directed at end users. Keep in mind that they probably do not care what file or class you changed, but what features you provided / what bugs you fixed. Try to answer the question "What benefit do I get from upgrading?"
More on http://keepachangelog.com

File: Resources/doc/cookbook/recipe_persisting_filters.rst [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+                arguments:
+                    - null
+                    - AppBundle\Entity\User
+                    - null
+
+
+.. note::
+
+    Both ``persist_filters`` and ``filter_persister`` can be used globally and per-admin.
+    Which provide you the most flexible way to configure this feature.
```
Comment (greg0ire) [inline Resources/doc/cookbook/recipe_persisting_filters.rst]: This last sentence looks a bit weird, because it's a separate sentence and it's unclear what you refer to.

File: Resources/doc/cookbook/recipe_persisting_filters.rst [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+Per Admin (using XML) :
+
+.. configuration-block::
+
+    .. code-block:: xml
+
+        <!-- src/AppBundle/Resources/config/admin.xml -->
+
+        <service id="app.admin.user" class="AppBundle\Admin\UserAdmin">
+            <tag name="sonata.admin" manager_type="orm" filters_persister="filter_persister_service_id" />
```
Comment (greg0ire) [inline Resources/doc/cookbook/recipe_persisting_filters.rst]: Should there really be an "s" in `filters_persister`?

Comment (greg0ire) [reply Resources/doc/cookbook/recipe_persisting_filters.rst]: There is no "s" on line 46

File: Filter/Persister/SessionFilterPersister.php [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Build the session key, under which the filter should be stored for given admin code.
+     *
+     * @param string $admin The admin code
+     *
+     * @return string The storage key
+     */
+    private function buildStorageKey($admin)
```
Comment (greg0ire) [inline Filter/Persister/SessionFilterPersister.php]: Please add return type hint

File: Filter/Persister/FilterPersisterInterface.php [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    public function set($admin, $filters);
+
+    /**
+     * Reset persisted filters for given admin.
+     *
+     * @param string $admin The admin code
+     */
+    public function reset($admin);
```
Comment (greg0ire) [inline Filter/Persister/FilterPersisterInterface.php]: Please add return type hinting to all methods

**Comment by yann-eugone** [comment]
@greg0ire ok, it took me some time, but I finally did found some time to push again on this. Let me tell if there is something else I can do (I was not sure about where to write the upgrade note).

File: Filter/Persister/FilterPersisterInterface.php [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+interface FilterPersisterInterface
+{
+    /**
+     * Get persisted filters for given admin.
+     *
+     * @param string $admin The admin code
+     *
+     * @return array The persisted filters
+     */
+    public function get(string $admin): array;
```
Comment (OskarStark) [inline Filter/Persister/FilterPersisterInterface.php]: imo $adminCode is more clear

File: UPGRADE-3.x.md [hunk 27]
```diff
@@ -1,6 +1,17 @@
 UPGRADE 3.x
 ===========
 
+UPGRADE FROM 3.32 to 3.33
```
Comment (OskarStark) [inline UPGRADE-3.x.md]: please remove headline, will be set after merge

**Comment by yann-eugone** [comment]
@greg0ire maybe i did not get your advice about adding typehint to all methods in the classes I've introduced. I did it, but as Travis testing matrix is including PHP < 7.0 cases, tests are failing...

**Comment by jordisala1991** [comment]
This PR started when we had 5.3 support, then we moved to 7.1 directly (that's when @greg0ire asked you to add typehints) but we rolled back to 5.6 .

So you can't add type hints for basic types or return types to functions. We have to support php 5.6

File: Resources/doc/cookbook/recipe_persisting_filters.rst [hunk 28]
```diff
@@ -0,0 +1,138 @@
+Persisting Filters
+==================
+
+Persisting filters allow your application to save the filters the authenticated user has submitted.
+Then the saved filters will be reused if the page is displayed again.
```
Comment (greg0ire) [inline Resources/doc/cookbook/recipe_persisting_filters.rst]: Please comply with the sf docs standards:

> [Each line should break approximately after the first word that crosses the 72nd character (so most lines end up being 72-78 characters);](http://symfony.com/doc/current/contributing/documentation/standards.html)

File: Resources/doc/cookbook/recipe_persisting_filters.rst [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+    .. code-block:: yaml
+
+        # app/config/config.yml
+
+        sonata_admin:
+            persist_filters: true
+            filter_persister: filter_persister_service_id
+
+
+Per Admin (using XML) :
```
Comment (jordisala1991) [inline Resources/doc/cookbook/recipe_persisting_filters.rst]: I think this can be simplified (because we use `configuration-block`) to:

```
.. configuration-block::

    .. code-block:: xml

        <!-- src/AppBundle/Resources/config/admin.xml -->

        <service id="app.admin.user" class="AppBundle\Admin\UserAdmin">
            <tag name="sonata.admin" manager_type="orm" filter_persister="filter_persister_service_id" />
            <argument />
            <argument>AppBundle\Entity\User</argument>
            <argument />
        </service>

    .. code-block:: yaml

        # src/AppBundle/Resources/config/admin.yml

        services:
            app.admin.user:
                class: AppBundle\Admin\UserAdmin
                tags:
                    -
                        name: sonata.admin
                        manager_type: orm
                        filter_persister: filter_persister_service_id
                arguments:
                    - null
                    - AppBundle\Entity\User
                    - null
```

There is also another of this simplifications, below.

**Comment by yann-eugone** [comment]
@greg0ire @OskarStark @jordisala1991 @core23 just rebased the whole Pull Request for the 2nd time, hope everything will be OK this time

**Comment by greg0ire** [comment]
Thanks a lot for this!

**Comment by OskarStark** [comment]
Thank you very much, great work @yann-eugone 👍

**Comment by yann-eugone** [comment]
Happy to finally find a way to contribute :smile:

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 30]
```diff
@@ -308,13 +308,20 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
 
         $definition->addMethodCall('setLabel', [$label]);
 
+        $persistFilters = $container->getParameter('sonata.admin.configuration.filters.persist');
+        // override default configuration with admin config if set
         if (isset($attributes['persist_filters'])) {
-            $persistFilters = (bool) $attributes['persist_filters'];
-        } else {
-            $persistFilters = (bool) $container->getParameter('sonata.admin.configuration.filters.persist');
+            $persistFilters = $attributes['persist_filters'];
```
Comment (jordisala1991) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Could be located here? @Soullivaneuh , before we had a cast to (bool) and now we don't. Maybe the value is getting there but not with the correct type?

Comment (kunicmarko20) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: well "true" would also be considered as a true so not sure that is it

Comment (jordisala1991) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: "true" !== true

Comment (kunicmarko20) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: if strict is used 😄

Comment (kunicmarko20) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: https://github.com/sonata-project/SonataAdminBundle/blob/0344393db83244b7422e75984c99a1bf7eb2de13/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L314-L323

and it is not
